benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,146 @@
+
+    _object_session = boto3.session.Session()
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    _branch: str = field(validator=instance_of(str), init=False)
+    """The branch of the BayBE library from which the workflow was started."""
+
+    _workflow_run_identifier: str = field(validator=instance_of(str), init=False)
+    """The identifier of the workflow run."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        ENVIRONMENT_NOT_SET = "BAYBE_BENCHMARK_PERSISTENCE_PATH" not in os.environ
+        if ENVIRONMENT_NOT_SET:
+            raise ValueError(
+                "The environment variable "
+                "BAYBE_BENCHMARK_PERSISTENCE_PATH is not set. "
+                "A bucket name must be provided."
+            )
+        return os.environ["BAYBE_BENCHMARK_PERSISTENCE_PATH"]
+
+    @_branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        S3_PATH_FORBIDDEN_CHARACTERS = ["/", "\\"]
+        sanitized_branch = ""
+        for char in S3_PATH_FORBIDDEN_CHARACTERS:
+            sanitized_branch = current_branch.replace(char, "-")
+        return sanitized_branch
+
+    @_workflow_run_identifier.default
+    def _default_workflow_id(self) -> str:
+        if "GITHUB_RUN_ID" not in os.environ:
+            raise ValueError("The environment variable GITHUB_RUN_ID is not set.")
+        return os.environ["GITHUB_RUN_ID"]
+
+    @override
+    def persist_new_result(self, result: Result, benchmark: Benchmark) -> None:
+        """Store the result of a benchmark.
+
+        This method will store the result of a benchmark in an S3 bucket.
+        The key of the object will be the experiment identifier, the branch, the
+        BayBE-version, the start datetime, the commit hash and the workflow
+        run identifier.
+
+        Args:
+            result: The result to be persisted.
+            benchmark: The benchmark definition that produced the result.
+        """
+        client = self._object_session.client("s3")
+
+        experiment_identifier = result.benchmark_identifier
+        metadata = result.metadata
+
+        bucket_path_key = (
+            f"{experiment_identifier}/{self._branch}/{metadata.latest_baybe_tag}/"
+            + f"{metadata.start_datetime.isoformat()}/"
+            + f"{metadata.commit_hash}/{self._workflow_run_identifier}"
+        )
+
+        result_dict = result.to_dict() | benchmark.to_dict()
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=f"{bucket_path_key}/result.json",
+            Body=json.dumps(result_dict),
+            ContentType="application/json",
+        )
+
+
+class LocalPersistence(PersistenceInterface):
+    """For saving results locally."""
+
+    def _build_filename(self, result: Result) -> str:
```

Comment:
I think you are making this overly complicated here: If people want to store results locally, they probably just want to give an explicit `Path` describing where stuff should be stored. So in my opinion, just having this class accept a `Path` as an attribute and doing some basic validation (does it already exist, is it a directory etc) makes more sense than us fixing a path. Of course, we can offer a sensible naming (basically what you do here already) as a default

Reply:
That should decently be doable . How about setting a default so that everyone can still execute `python -m benchmarks` which results in an autogenerated name but if the module is used, a name can be set manually?

Reply:
Looks good to me now, only issue I see at the moment is that stuff should be a `Path`, but since there is another comment for that, I'll resolve here.